File: nodejs-typescript/src/application/RollingBackLedger.ts
import type { CompanyAccountBook } from '../domain/CompanyAccountBook.js';
import type { Ledger } from '../domain/Ledger.js';

/** Synchronous transactional boundary: on any error during apply, restore pre-transfer balances. */
export class RollingBackLedger implements Ledger {
  constructor(private readonly accountBook: CompanyAccountBook) {}

  atomic<T>(operation: () => T): T {
    const snapshot = this.accountBook.captureBalanceSnapshot();
    try {
      return operation();
    } catch (error) {
      this.accountBook.restoreBalanceSnapshot(snapshot);
      throw error;
    }
  }
}

File: ruby/lib/mable_bank/application/rolling_back_ledger.rb
# frozen_string_literal: true

module MableBank
  module Application
    # Synchronous transactional boundary: on any error during apply, restore pre-transfer balances.
    class RollingBackLedger
      include Ledger

      def initialize(account_book)
        @account_book = account_book
      end

      def atomic
        snapshot = @account_book.capture_balance_snapshot
        yield
      rescue StandardError
        @account_book.restore_balance_snapshot(snapshot)
        raise
      end
    end
  end
end
